### ISSUE TITLE
sympy test failure

### ISSUE THREAD

**Issue by verdurin** [hpcugent/easybuild-easyconfigs#2037]
I'm working on porting sympy-0.16.1 for goolf 1.7.20.

The build seems to be going well until the test phase, when there is an error:

```
sympy/external/tests/test_autowrap.py[0] Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "sympy/utilities/runtests.py", line 516, in _test
    slow=slow, enhance_asserts=enhance_asserts))
  File "sympy/utilities/runtests.py", line 981, in test
    self.test_file(f, sort, timeout, slow, enhance_asserts)
  File "sympy/utilities/runtests.py", line 1046, in test_file
    exec_(code, gl)
  File "sympy/core/compatibility.py", line 137, in exec_
    exec("exec _code_ in _globs_, _locs_")
  File "<string>", line 1, in <module>
  File "/tmp/centos/sympy/0.7.6.1/goolf-1.7.20-Python-2.7.9/sympy-sympy-0.7.6.1/sympy/external/tests/test_autowrap.py", line 14, in <module>
    autowrap(symbols('x'), 'f95', 'f2py')
  File "sympy/core/cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
  File "sympy/core/compatibility.py", line 872, in wrapper
    result = user_function(*args, **kwds)
  File "sympy/utilities/autowrap.py", line 508, in autowrap
    return code_wrapper.wrap_code(routine, helpers=helps)
  File "sympy/utilities/autowrap.py", line 144, in wrap_code
    self._process_files(routine)
  File "sympy/utilities/autowrap.py", line 163, in _process_files
    " ".join(command), e.output.decode()))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 4969: ordinal not in range(128)
```

The test process then restarts and the remaining tests either succeed or are skipped:

```
============================= test process starts ==============================
executable:         /opt/soft/apps/Python/2.7.9-goolf-1.7.20/bin/python  (2.7.9-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python
hash randomization: on (PYTHONHASHSEED=592645159)

sympy/vector/vector.py[9] .........                                         [OK]
```

etc.

```
========== tests finished: 2882 passed, 14 skipped, in 249.02 seconds ==========
```
and it ends like this:

```
(at easybuild/tools/run.py:360 in parse_cmd_output)
== 2015-10-07 10:57:40,160 main.easyblock WARNING build failed (first 300 chars): cmd "python setup.py test" exited with exitcode 1 and output:
running test
============================= test process starts ==============================
executable:         /opt/soft/apps/Python/2.7.9-goolf-1.7.20/bin/python  (2.7.9-final-0) [CPython]
architecture:       64-bit
cache:
== 2015-10-07 10:57:40,161 main.PythonPackage INFO Closing log for application name sympy version 0.7.6.1
== 2015-10-07 10:57:40,161 main.easyblock INFO FAILED: Installation ended unsuccessfully (build directory: /tmp/centos/sympy/0.7.6.1/goolf-1.7.20-Python-2.7.9): build failed (first 300 chars): cmd "python setup.py test" exited with exitcode 1 and output:
running test
============================= test process starts ==============================
executable:         /opt/soft/apps/Python/2.7.9-goolf-1.7.20/bin/python  (2.7.9-final-0) [CPython]
architecture:       64-bit
cache:
== 2015-10-07 10:57:40,161 main.easyblock INFO Results of the build can be found in the log file /tmp/centos/easybuild-sympy-0.7.6.1-20151007.103926.CePkl.log
== 2015-10-07 10:57:40,162 main ERROR build failed (first 300 chars): cmd "python setup.py test" exited with exitcode 1 and output:
running test
============================= test process starts ==============================
executable:         /opt/soft/apps/Python/2.7.9-goolf-1.7.20/bin/python  (2.7.9-final-0) [CPython]
architecture:       64-bit
cache:             (at easybuild/main.py:115 in build_and_install_software)
== 2015-10-07 10:57:40,163 main.filetools DEBUG Not creating existing path /tmp/centos
== 2015-10-07 10:57:40,163 main ERROR Build of /home/centos/src/easybuild-easyconfigs-devel/sympy-0.7.6.1-goolf-1.7.20-Python-2.7.9.eb failed (err: 'build failed (first 300 chars): cmd "python setup.py test" exited with exitcode 1 and output:\nrunning test\n============================= test process starts ==============================\nexecutable:         /opt/soft/apps/Python/2.7.9-goolf-1.7.20/bin/python  (2.7.9-final-0) [CPython]\narchitecture:       64-bit\ncache:            ') (at easybuild/main.py:147 in build_and_install_software)
```

**Comment by boegel** [comment]
This is probably caused by a combination of a bug in the particular Python version you're using, and your locale config ($LANG, $LC_ALL), etc.
See also https://bugs.launchpad.net/pyskein/+bug/709247, for example.

I have this in my .bashrc, partially because of this problem:

```
export LANG=en_US.utf8
export LC_ALL=en_US.utf8
```

Let us know whether this helps.

**Comment by verdurin** [comment]
No, that didn't fix it:

```
[centos@easybuild-6 easybuild-easyconfigs-devel]$ echo $LANG $LC_ALL
en_US.utf8 en_US.utf8
```

and exactly the same crash happened during the tests. Hmmph.

**Comment by verdurin** [comment]
A minor update in that I've tried this again with foss-2015b and the same thing happens, even with the locale change.

**Comment by boegel** [comment]
@verdurin any updates on this?

**Comment by verdurin** [comment]
Tests still fail with `2.6.0`. 

I will update the EB to the recent `1.0.0` release and try that.

**Comment by verdurin** [comment]
This fails with an updated easyconfig for `sympy-1.0`.

Will try with the locale hack.

**Comment by verdurin** [comment]
Okay, that attempt failed too (with `$LANG` and `$LC_ALL` set to `en_US.utf8`.

**Comment by verdurin** [comment]
Upstream bug report at:

https://github.com/sympy/sympy/issues/10818

**Comment by dominikborkowski** [comment]
Just to add more data points, this error also occurs on foss-2016a with Python 2.7.11 and sympy 0.7.6:

sympy/external/tests/test_autowrap.py[0] Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "sympy/utilities/runtests.py", line 516, in _test
    slow=slow, enhance_asserts=enhance_asserts))
  File "sympy/utilities/runtests.py", line 981, in test
    self.test_file(f, sort, timeout, slow, enhance_asserts)
  File "sympy/utilities/runtests.py", line 1046, in test_file
    exec_(code, gl)
  File "sympy/core/compatibility.py", line 137, in exec_
    exec("exec _code_ in _globs_, _locs_")
  File "<string>", line 1, in <module>
  File "/dev/shm/dom/build/sympy/0.7.6/foss-2016a-Python-2.7.11/sympy-0.7.6/sympy/external/tests/test_autowrap.py", line 14, in <module>
    autowrap(symbols('x'), 'f95', 'f2py')
  File "sympy/core/cache.py", line 91, in wrapper
    retval = cfunc(*args, **kwargs)
  File "sympy/core/compatibility.py", line 872, in wrapper
    result = user_function(*args, **kwds)
  File "sympy/utilities/autowrap.py", line 521, in autowrap
    return code_wrapper.wrap_code(routine, helpers=helps)
  File "sympy/utilities/autowrap.py", line 144, in wrap_code
    self._process_files(routine)
  File "sympy/utilities/autowrap.py", line 163, in _process_files
    " ".join(command), e.output.decode()))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 4163: ordinal not in range(128)
============================= test process starts ==============================
executable:         /apps/easybuild/software/shadowfax-westmere/Python/2.7.11-foss-2016a/bin/python  (2.7.11-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python
hash randomization: on (PYTHONHASHSEED=2567096024)

**Comment by boegel** [comment]
@verdurin, @dominikborkowski Can you check whether patching `sympy/utilities/autowrap.py:163` to be `" ".join(command), e.output.decode('ascii', 'ignore')))` fixes the issue?

See http://stackoverflow.com/questions/21129020/how-to-fix-unicodedecodeerror-ascii-codec-cant-decode-byte and http://nedbatchelder.com/text/unipain/unipain.html for detailed information about the unicode mess in Python 2.x.

**Comment by boegel** [comment]
@verdurin can this be closed now #3058 is merged?

**Comment by verdurin** [comment]
Yes. Closed.
